DPGAnalysis/L1TNanoAOD/plugins/CaloTPTableProducer.cc

```diff
@@ -0,0 +1,258 @@
+    for (auto itr : (*hcalTPs.product())) {
+      int ver = itr.id().version();
+      short ieta = (short)itr.id().ieta();
+      unsigned short absIeta = (unsigned short)abs(ieta);
+      unsigned short cal_iphi = (unsigned short)itr.id().iphi();
+      unsigned short iphi = (72 + 18 - cal_iphi) % 72;
+
+      unsigned short compEt = itr.SOI_compressedEt();
+      double et = decoder->hcaletValue(itr.id(), itr.t0());
+      unsigned short fineGrain = (unsigned short)itr.SOI_fineGrain();
+
+      if (compEt > 0 && (absIeta < 29 || ver == 1)) {
+        hcalTPieta.push_back(ieta);
+        hcalTPCaliphi.push_back(cal_iphi);
+        hcalTPiphi.push_back(iphi);
+        hcalTPet.push_back(et);
+        hcalTPcompEt.push_back(compEt);
+        hcalTPfineGrain.push_back(fineGrain);
+        nHCALTP++;
+      }
+    }
+  }
+
+  auto hcalTPTable = std::make_unique<nanoaod::FlatTable>(nHCALTP, hcalTPsName_, false);
+  hcalTPTable->addColumn<int16_t>("ieta", hcalTPieta, "");
+  hcalTPTable->addColumn<int16_t>("Caliphi", hcalTPCaliphi, "");
+  hcalTPTable->addColumn<int16_t>("iphi", hcalTPiphi, "");
+  hcalTPTable->addColumn<float>("et", hcalTPet, "", 12);
+  hcalTPTable->addColumn<int16_t>("compEt", hcalTPcompEt, "");
+  hcalTPTable->addColumn<int16_t>("fineGrain", hcalTPfineGrain, "");
+
+  iEvent.put(std::move(ecalTPTable), "HcalTP");
+  iEvent.put(std::move(hcalTPTable), "EcalTP");
+}
+
+// ------------ method called once each stream before processing any runs, lumis or events  ------------
+void CaloTPTableProducer::beginStream(edm::StreamID) {
+  // please remove this method if not needed
+}
+
+// ------------ method called once each stream after processing all runs, lumis and events  ------------
+void CaloTPTableProducer::endStream() {
+  // please remove this method if not needed
+}
+
+// ------------ method called when starting to processes a run  ------------
+/*
+void
+CaloTPTableProducer::beginRun(edm::Run const&, edm::EventSetup const&)
+{
+}
+*/
+
+// ------------ method called when ending the processing of a run  ------------
+/*
+void
+CaloTPTableProducer::endRun(edm::Run const&, edm::EventSetup const&)
+{
+}
+*/
+
+// ------------ method called when starting to processes a luminosity block  ------------
+/*
+void
+CaloTPTableProducer::beginLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&)
+{
+}
+*/
+
+// ------------ method called when ending the processing of a luminosity block  ------------
+/*
+void
+CaloTPTableProducer::endLuminosityBlock(edm::LuminosityBlock const&, edm::EventSetup const&)
+{
+}
+*/
+
+// ------------ method fills 'descriptions' with the allowed parameters for the module  ------------
+void CaloTPTableProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  //The following says we do not know what parameters are allowed so do no validation
+  // Please change this to state exactly what you do use, even if it is no parameters
+  edm::ParameterSetDescription desc;
```

Comment:
please add the proper parameter validation : see https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp